### PR TITLE
[Fusion] Manual `ActivationQuantFusionPass` initial application

### PR DESCRIPTION
**Pull request by mgoin**
Starts the `ActivationQuantFusionPass` manual-fusion migration (RFC #43224, specific tracker https://github.com/vllm-project/vllm/issues/43501) on the producer side of the `QuantizedActivation` contract (#44260). 

Adds `maybe_fused_act_quant`: given an activation and the linear it feeds, it emits a `QuantizedActivation` via the fused `silu_and_mul_quant` kernel when the linear advertises a consumable `input_quant_key`, and falls back to the plain activation otherwise. Llama's MLP forward calls it for `down_proj` — the model-code change is one line. When manual fusion fires the `silu_and_mul` pattern is already consumed, so the compiler `ActivationQuantFusionPass` finds nothing and the two never double-fuse. Only `(SiluAndMul, kFp8StaticTensorSym)` is registered for now; other schemes/activations are one table row each.

Not a duplicate: builds on the landed contract (#44260) and is the activation-quant counterpart to the AR+RMSNorm manual fusion in #45855; no open PR covers the producer side.

## Test
Smoke tested on `RedHatAI/Llama-3.2-1B-Instruct-FP8` (default compiled + CUDA-graph path): all 16 `mlp.down_proj` layers take the fused branch (`input_quant_key == kFp8StaticTensorSym`) and generations are coherent — the `QuantizedActivation` flows through the compiled `down_proj.forward` into `apply_weights` with no graph break. TP2 coherence and an A/B vs. the compiler pass are follow-ups.

AI assistance (Claude) was used; all changed lines are human-reviewed.

### CODE REVIEW

File: vllm/model_executor/layers/fusion/fused_act_quant.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    return QuantizedActivation(
+        data=result,
+        scale=scale,
+        orig_dtype=x.dtype,
+        orig_shape=out_shape,
+        quant_key=kFp8StaticTensorSym,
+    )
+
+
+# (activation module type, consumer input_quant_key) -> fused producer.
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fusion/fused_act_quant.py:131]: do you think we should build in some oracle for this selection up front where the actual implementation of the quantized activation function can be swapped?

File: vllm/model_executor/models/llama.py [hunk 2]
```diff
@@ -115,7 +116,7 @@ def __init__(
 
     def forward(self, x):
         x, _ = self.gate_up_proj(x)
-        x = self.act_fn(x)
+        x = maybe_fused_act_quant(self.act_fn, x, self.down_proj)
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/models/llama.py:118]: I think that instead of passing `self.down_proj: LinearBase`, it might make more sense to expose a method from `LinearBase` that can be called here to extract what is needed

I think that will help make the interface clearer

**Comment by whkuo269** [comment]
Hi @mgoin, I’d like to help validate this manual activation+quant fusion work.

I’m planning to run TP=1/TP=2 correctness checks and A/B tests comparing the base commit and this PR with compiler activation-quant fusion enabled and disabled. 

Please let me know if there’s a preferred test setup or specific coverage you’d find most useful.

File: vllm/model_executor/models/llama.py (hunk 2 above)
Comment (mgoin) [reply vllm/model_executor/models/llama.py:118]: Depending on the quantization used, we need to extract different things from the Linear layer. IMO it is more clear to my eye that this function might have an affect on `self.down_proj` within the context of the forward function, so I prefer just passing it in

File: vllm/model_executor/layers/fusion/fused_act_quant.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/layers/fusion/fused_act_quant.py:131]: Yeah it is reasonable to suspect for future optimizations

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by njhill** [comment]
@mgoin I bisected a CI IMA to this change: https://buildkite.com/vllm/ci/builds/87269/list?jid=01a06dce-0d6b-4b62-8db0-9853df2b22d9&tab=output

**Comment by rishabhsinha17** [comment]
Taking the A/B vs the compiler pass that the description lists as a follow-up, per the Slack thread with @mgoin.

Same commit, two arms. Arm A is main as merged, where maybe_fused_act_quant fires and ActivationQuantFusionPass finds nothing to match. Arm B reverts the one llama.py line to self.act_fn(x) and forces the silu_and_mul and quant_fp8 custom ops on, mirroring the tests/compile/fusions_e2e configs, so the pass does the fusing as before this PR. Each arm gets a fusion check first, match_table count on the pass side and a fused-branch counter on the manual side, then decode TPOT and throughput sweeps plus gsm8k parity on RedHatAI/Llama-3.2-1B-Instruct-FP8 and RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8. Hardware is one L40S, which covers the FP8 static per-tensor row. NVFP4 needs SM100 and TP2 a second GPU, so I am leaving those to a later run. I will add an eager -O0 datapoint since the manual path still fuses there while the pass needs compilation. Results here within a week, then model batch claims on #43501.

@whkuo269 you offered validation on this thread pre-merge, happy to coordinate if you still have WIP.

**Comment by whkuo269** [comment]
@rishabhsinha17  Thanks for circling back! I don’t have an active WIP at the moment, but I’d be happy to collaborate on the validation. I can help with any remaining coverage or independently reproduce part of the matrix. If you share the exact commit, Arm B diff, and benchmark commands/configs, I can keep my run aligned with yours.

**Comment by rishabhsinha17** [comment]
A/B results are in, manual fusion vs the compiler pass on the same commit. Short version. Both paths end in the same `_C.silu_and_mul_quant` kernel and they are equal end to end within about one percent on both models. The only visible gap comes from force-enabling the `silu_and_mul` and `quant_fp8` custom ops, which the pass path needs to see its pattern and the manual path does not need at all. Accuracy is identical within noise. The manual path keeps fusing under `-O0` eager, out of the pass's reach. This supports keeping the manual path as the default owner of this boundary.

Setup. Commit a6131b0e656e18eeb3797b8856aa3dbc69efd8e5 for every arm, one L40S (SM89), serial runs, host otherwise idle. Models RedHatAI/Llama-3.2-1B-Instruct-FP8 (16 layers) and RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8 (32 layers), both FP8 static per-tensor activation. Arm A is main as merged (manual fusion). Arm B reverts the one llama.py line and forces `-cc '{"custom_ops":["+silu_and_mul","+quant_fp8"],"pass_config":{"fuse_act_quant":true}}'`, the pre-#51415 CI shape. To isolate the op-forcing confound, arm A-ops runs the merged tree under those same forced flags. Arm C turns both fusers off, and arm D is `-O0`.

I gated fusion engagement before taking any perf number, using the `get_compilation_match_table` worker RPC plus the manual-side counter from #55338 (ElizaWszola's two hunks, applied only during verification and reverted before every benchmark, credit to her).

| model | arm | manual count | pass matches |
|---|---|---|---|
| 1b | A | 16 | 0 |
| 1b | A, pass force-enabled | 16 | 0 |
| 1b | A-ops | 16 | 0 |
| 1b | B | 0 | 16 |
| 1b | C | 0 | 0 |
| 1b | D (eager) | 336 (16 x 21 forwards) | 0 |
| 8b | A | 32 | 0 |
| 8b | B | 0 | 32 |
| 8b | D (eager) | 672 | 0 |

On the merged tree the force-enabled pass executes and its match count stays at 0, while the manual counter reads n_layers. On the reverted tree the pass matches n_layers. In the arm B fx graphs the pre-pass unit holds `_C.silu_and_mul` feeding `_C.static_scaled_fp8_quant` and the post-pass unit holds `_C.silu_and_mul_quant`. Arm A's pre-pass graph already holds `_C.silu_and_mul_quant`. The pass has no pattern left to find. One reproducibility note. The match counter is cache-regime dependent. With warm inductor caches the 16 identical layer submodules deduplicate and the pass reports 2 (one per unique compiled unit) instead of 16, so run any match-count check with compile caches disabled or audit the dumped graphs. My gate does both (all n_layers boundaries present pre-split, zero unfused `silu_and_mul` in any final unit).

Decode latency, `vllm bench latency`, input 512, output 128, 10 iters after 3 warmup, ms per output token, stdev in parens.

1B:

| bs | A | A-ops | B | C | D | B vs A % |
|---|---|---|---|---|---|---|
| 1 | 2.947 (0.002) | 2.987 (0.002) | 2.990 (0.002) | 2.944 (0.034) | 10.285 (0.072) | +1.45 |
| 2 | 3.059 (0.017) | 3.100 (0.016) | 3.090 (0.013) | 3.030 (0.021) | 10.646 (0.051) | +1.00 |
| 4 | 3.231 (0.027) | 3.214 (0.003) | 3.243 (0.027) | 3.176 (0.033) | 10.646 (0.060) | +0.36 |
| 8 | 3.430 (0.005) | 3.483 (0.022) | 3.437 (0.002) | 3.424 (0.016) | 10.872 (0.104) | +0.19 |
| 16 | 4.058 (0.029) | 4.075 (0.015) | 4.062 (0.004) | 4.015 (0.025) | 11.233 (0.062) | +0.10 |
| 32 | 5.495 (0.019) | 5.495 (0.034) | 5.518 (0.012) | 5.425 (0.012) | 11.940 (0.037) | +0.42 |
| 64 | 7.851 (0.019) | 7.870 (0.031) | 7.879 (0.017) | 7.773 (0.011) | 14.107 (0.056) | +0.36 |
| 128 | 12.846 (0.051) | 12.950 (0.028) | 12.936 (0.025) | 12.836 (0.041) | 17.509 (0.076) | +0.70 |
| 256 | 22.466 (0.080) | 22.763 (0.064) | 22.751 (0.103) | 22.416 (0.109) | 24.856 (0.128) | +1.27 |

8B:

| bs | A | B | D | B vs A % |
|---|---|---|---|---|
| 1 | 14.386 (0.003) | 14.385 (0.008) | 19.113 (0.074) | -0.00 |
| 2 | 14.811 (0.046) | 14.836 (0.006) | 19.439 (0.085) | +0.17 |
| 4 | 15.462 (0.014) | 15.442 (0.011) | 20.232 (0.047) | -0.14 |
| 8 | 16.820 (0.036) | 16.855 (0.007) | 20.782 (0.095) | +0.21 |
| 16 | 19.501 (0.017) | 19.584 (0.026) | 22.955 (0.084) | +0.43 |
| 32 | 25.503 (0.046) | 25.510 (0.061) | 26.815 (0.122) | +0.03 |
| 64 | 37.031 (0.119) | 37.187 (0.123) | 38.024 (0.082) | +0.42 |
| 128 | 60.179 (0.251) | 60.353 (0.132) | 61.478 (0.232) | +0.29 |
| 256 | 111.014 (4.806) | 109.099 (0.431) | 110.683 (0.651) | -1.73 |

The 8B bs256 A row has one outlier iteration (stdev 4.8 vs 0.4 elsewhere), so treat that -1.73 as noise from the outlier. The clean decomposition on the 1B at bs1 reads A 2.947, A-ops 2.987, B 2.990. Forcing the custom ops costs +1.36% and pass-vs-manual on identical op forms costs +0.10% on top. C tracks A within noise to -1.3%, so at 1B scale inductor's native codegen of this boundary is already competitive end to end, and the kernel-level win below mostly disappears into the runtime share of this boundary.

Throughput, `vllm bench throughput`, random dataset, tok/s.

| model | arm | 512x128 | 2048x256 |
|---|---|---|---|
| 1b | A | 56567 | 57675 |
| 1b | A-ops | 56134 | 57170 |
| 1b | B | 56118 | 57120 |
| 1b | C | 56584 | 57563 |
| 8b | A | 10090 | 10333 |
| 8b | B | 10099 | 10358 |

B vs A on the 1B is -0.79% and -0.96%, and B vs A-ops is -0.03% and -0.09%, same attribution as latency. The 8B pair is even within noise.

gsm8k, lm_eval local-completions, 5-shot, all 1319 questions, strict exact match with stderr.

| model | A | B |
|---|---|---|
| 1b | 0.3321 (0.0130) | 0.3328 (0.0130) |
| 8b | 0.6998 (0.0126) | 0.7043 (0.0126) |

Kernel microbench for the static per-tensor row (neither in-repo silu bench covers it), unfused `_C.silu_and_mul` + `_C.static_scaled_fp8_quant` vs fused `_C.silu_and_mul_quant`, bf16 to fp8e4m3, all rows bitwise equal.

| d | tokens | unfused us | fused us | speedup |
|---|---|---|---|---|
| 8192 | 1 | 12.23 | 6.05 | x2.02 |
| 8192 | 256 | 12.36 | 7.82 | x1.58 |
| 14336 | 128 | 12.25 | 7.92 | x1.55 |
| 14336 | 4096 | 776.6 | 408.7 | x1.90 |
| 14336 | 16384 | 3114.7 | 1621.1 | x1.92 |

@whkuo269 thanks for the offer to reproduce, everything needed is below and I am happy to compare notes on any slice of the matrix.

<details>
<summary>Exact commit, arm B diff, and commands</summary>

Commit `a6131b0e656e18eeb3797b8856aa3dbc69efd8e5`, install with `VLLM_USE_PRECOMPILED=1 pip install -e .` (torch 2.13.0+cu130 resolves).

Arm B tree change (revert of the merged one-liner):

```diff
--- a/vllm/model_executor/models/llama.py
+++ b/vllm/model_executor/models/llama.py
@@ -117,7 +117,7 @@ class LlamaMLP(nn.Module):

     def forward(self, x):
         x, _ = self.gate_up_proj(x)
-        x = maybe_fused_act_quant(self.act_fn, x, self.down_proj)
+        x = self.act_fn(x)
         x, _ = self.down_proj(x)
         return x
```

Arm flags. A none, A-ops and B `-cc "$CC"` with

```bash
CC='{"custom_ops":["+silu_and_mul","+quant_fp8"],"pass_config":{"fuse_act_quant":true}}'
```

C `-cc '{"pass_config":{"fuse_act_quant":false}}'` on the arm B tree, D `--optimization-level 0`.

Fusion check (run with caches off so the passes execute, see the note above):

```bash
export VLLM_DISABLE_COMPILE_CACHE=1 VLLM_CACHE_ROOT=$(mktemp -d) TORCHINDUCTOR_FORCE_DISABLE_CACHES=1
python - <<'PY'
from vllm import LLM, SamplingParams
llm = LLM(model="RedHatAI/Llama-3.2-1B-Instruct-FP8", max_model_len=2048,
          # per arm: compilation_config={...} / optimization_level=0
          )
llm.generate(["The capital of France is"], SamplingParams(temperature=0, max_tokens=16))
print(llm.llm_engine.engine_core.collective_rpc("get_compilation_match_table"))
PY
```

The pass side reads `activation_quant_fusion_pass` from that table in-tree. The manual-side counter needs the two hunks from #55338 (`fused_act_quant.py` counter plus the `gpu_worker.py` RPC line).

Benchmarks per arm and model:

```bash
M=RedHatAI/Llama-3.2-1B-Instruct-FP8   # and the 8B
for BS in 1 2 4 8 16 32 64 128 256; do
  vllm bench latency --model $M ${ARM_FLAGS[@]} --max-model-len 4096 \
    --gpu-memory-utilization 0.90 --input-len 512 --output-len 128 \
    --batch-size $BS --num-iters-warmup 3 --num-iters 10 --output-json lat_$BS.json
done
vllm bench throughput --model $M ${ARM_FLAGS[@]} --max-model-len 4096 \
  --gpu-memory-utilization 0.90 --dataset-name random --input-len 512 \
  --output-len 128 --num-prompts 512 --output-json tp_512x128.json
vllm bench throughput --model $M ${ARM_FLAGS[@]} --max-model-len 4096 \
  --gpu-memory-utilization 0.90 --dataset-name random --input-len 2048 \
  --output-len 256 --num-prompts 256 --output-json tp_2048x256.json
vllm serve $M ${ARM_FLAGS[@]} --port 8321 --max-model-len 4096 --gpu-memory-utilization 0.90 &
lm_eval --model local-completions --model_args \
  "model=$M,base_url=http://127.0.0.1:8321/v1/completions,num_concurrent=64,max_retries=3,tokenized_requests=False" \
  --tasks gsm8k --num_fewshot 5 --batch_size 1
```

</details>

Limitations. One L40S, so this covers the FP8 static per-tensor row only. NVFP4 needs SM100 and TP2 a second GPU, both out of scope here. #55338 adds TP2 CI coverage. Happy to run the same harness on the #43501 model batches as they land.
